osbuild/solver/pacman.py

```diff
@@ -0,0 +1,143 @@
+    lines = text.split("\n")
+
+    def parse_line(l):
+        k, v = l.split(":", maxsplit=1)
+        return k.strip(), v.strip()
+
+    return dict(parse_line(line) for line in lines if ":" in line)
+
+
+class Pacman(SolverBase):
+    def __init__(self, request: dict, cache_path: str) -> None:
+        self._cache_path = pathlib.Path(cache_path)
+
+        self._etc_path = self._cache_path / "etc"
+        self._cfg_path = self._etc_path / "pacman.conf"
+
+        self._lib_path = self._cache_path / "var/lib/pacman"
+
+        self._architecture = request["arch"]
+        self._repositories = request.get("arguments", {}).get("repos", [])
+
+    def _prepare_root(self) -> None:
+        """Prepares a pacman root for a our architecture by preparing the
+        necessary directories and writing a pacman configuration file that
+        includes the configured repositories."""
+
+        self._etc_path.mkdir(parents=True, exist_ok=True)
+        self._lib_path.mkdir(parents=True, exist_ok=True)
+
+        text = CONFIGURATION_TEMPLATE.format(architecture=self._architecture)
+
+        for repository in self._repositories:
+            text = text + REPOSITORY_TEMPLATE.format(
+                id=repository["id"], url=repository["baseurl"]
+            )
+
+        self._cfg_path.write_text(text)
+
+    def depsolve(self, arguments: dict) -> dict:
+        self._prepare_root()
+
+        # Ensure there is exactly one transaction given by unpacking the given
+        # transactions into a single name
+        (transaction,) = arguments.get("transactions", [])
+
+        # Run pacman's sync and refresh first to make sure the root is valid
+        subprocess.run(
+            [
+                "pacman", "-Sy",
+                "--root", str(self._cache_path),
+                "--config", str(self._cfg_path),
+            ],
+            stdout=subprocess.DEVNULL,
+            stderr=subprocess.DEVNULL,
+            check=True,
+        )
+
+        # Pacman does not have any form of JSON output so we're going to build
+        # JSON with string formatting!
+        result = subprocess.run(
+            [
+                "pacman", "-S",
+                "--print",
+                "--print-format", PRINT_FORMAT,
+                "--sysroot", str(self._cache_path),
+                *transaction["package-specs"]["include"],
+            ],
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            check=True,
+        ).stdout.decode("utf-8")
+
+        result = result.strip().rstrip(",")
+        result = f"[{result}]"
+
+        # The certainty of this being valid JSON is: "uhh, maybe" so let's try
+        # to load it.
+        selected = json.loads(result)
+        packages = []
+
+        for selection in selected:
+            # We need to request more information from pacman for each package
+            text = subprocess.run(
+                [
+                    "pacman", "-Sii",
```

Comment:
And yeah this "works" but there is a slight oversight in the buildtime being printed in the users locale. It should probably be epoch

```
[jelle@natrium][~]%pacman -S --print --print-format '{ "url": "%u", "arch": "%a", "name": "%n", "license": "%L", "buildtime": "%b", "description": "%d" }' linux
{ "url": "https://github.com/archlinux/linux", "arch": "x86_64", "name": "linux", "license": "GPL-2.0-only", "buildtime": "di 01 okt 2024 01:49:50 CEST", "description": "The Linux kernel and modules" }
```

Reply:
And reported https://gitlab.archlinux.org/pacman/pacman/-/issues/198

Reply:
Thank you, we can probably assume pacman 7.0 since it's in Fedora 41, perhaps I'll add a version check in the depsolver and spec file for the possible subpackage.